### DISCUSSION TITLE
Problem with next/script and next/link

### DISCUSSION THREAD

**Discussion by itaquito** [vercel/next.js#38048] [Help] [answered]
Hello!

I'm trying to add a Twitch player to my website. Adding a Twitch player to a website is relatively easy, you just need to load this script `https://embed.twitch.tv/embed/v1.js` and then initialize an object:
```js
new Twitch.Embed("twitch-embed", {
        width: 854,
        height: 480,
        channel: "monstercat",
        // Only needed if this page is going to be embedded on other websites
        parent: ["embed.example.com", "othersite.example.com"]
      });
```

So I loaded the script with `next/script` and added that piece of code into the onLoad function:
```js
<Script
    src='https://embed.twitch.tv/embed/v1.js'
    strategy='lazyOnload'
    onLoad={() => {
        new Twitch.Embed('twitch', {
            width: 427,
            height: 240,
            channel: 'bobicraftmc',
            layout: 'video',
            allowfullscreen: true
        });
    }}
/>
```

Everything works fine, the player shows correctly on my website. The problem appears when I use `next/link`. When I go to another page via the `<Link>` component and then come back to the page containing the script, again, via the `<Link>` component, the onLoad function doesn't execute again causing the player to not appear. I guess it's because the script remains loaded.

So, does anyone know is there a way to fix this issue?

Thanks in advance

**Comment by chris-m-taylor** [comment]
I have the same problem.

**Comment by ryanhaticus** [comment, marked as answer]
The issue you're facing is actually related to the browser itself. As your NextJS application is an SPAs (meaning no "hard refreshes" are ever completed), the browser never actually eradicates the script when the page is changed. There are two viable solutions to this issue, one good, one not so good.

Firstly, the not so good option:
1. Reloading the page whenever you don't need the script anymore.
This can be done by listening to the `router.asPath` in a `useEffect`, but it certainly isn't ideal and will create a poor user experience.

And finally, the good option:
1. Create a React Context with a boolean variable that stores whether or not the script has already been loaded. If it has, then call the Twitch SDK as per usual in a `useEffect` or something similar.

For example, I may have the `TwitchProvider.tsx` which contains context and the actual provider component:
```jsx
const TwitchContext = React.createContext({ loaded: false, setLoaded: null });

export const useTwitch = () => React.useContext(TwitchContext);

const TwitchProvider = ({ children }) => {
  const [loaded, setLoaded] = React.useState(false);

  return <TwitchContext.Provider value={{loaded, setLoaded}}>{children}</TwitchContext.Provider>
}

export default TwitchProvider;
```

Then, in my `_app.js` file, I will wrap my component as such:
```jsx
import TwitchProvider from './TwitchProvider'

const App = ({ Component, pageProps }) => {
  return (
    <TwitchProvider>
        <Component {...pageProps} />
    </TwitchProvider>
  );
};

export default App;
```

Finally, in my `component.js` file, I will use a `useEffect` to determine whether or not I should load a new embed, etc:
```jsx
import { useTwitch } from './TwitchProvider'

const Component = () => {
  const { loaded, setLoaded } = useTwitch();
  const router = useRouter(); // from next/router

  useEffect(() => {
    if (!loaded) {
      return;
    }

    // Change this to your page's path, as needed.
    if (router.asPath !== 'document') {
      return;
    }

    loadEmbed();
  }, [loaded, router.asPath])

  const loadEmbed = () => {
    new Twitch.Embed('twitch', {
      width: 427,
      height: 240,
      channel: 'bobicraftmc',
      layout: 'video',
      allowfullscreen: true
    });
  }

  return <>
    <Script src='https://embed.twitch.tv/embed/v1.js'
      strategy='lazyOnload'
      onLoad={() => {
        setLoaded(true);
      }}
    />
  </>
}
```

This is the kind of approach I would use as a work around. Please note, I didn't test this code, it's just an example. For more complete documentation regarding context, please see [the documentation](https://reactjs.org/docs/context.html).

**Comment by itaquito** [comment]
I actually managed to solve this issue sometime after I posted this thread.
First, I created a state that changes when the script loads the **first time**:
```js
const [isTwitchLoaded, setTwitchLoaded] = useState();
```
and
```js
<Script
    src='https://embed.twitch.tv/embed/v1.js'
    strategy='lazyOnload'
    onLoad={() => {
        setTwitchLoaded(true);
    }}
/>
```

Then, I use `useEffect` to load my player:
```js
useEffect(() => {
        if (isTwitchLoaded || typeof Twitch !== 'undefined') {
            new Twitch.Embed('twitch', {
                width: 640,
                height: 360,
                channel: 'bobicraftmc',
                layout: 'video',
                allowfullscreen: true
            });
        }
    }, [isTwitchLoaded]);
   ```
You probably noticed that I did `typeof Twitch !== 'undefined'`. This also loads the player if the `Twitch` variable is defined (meaning that the script is loaded). Probably, this method won't work with all the scripts out there, since a lot of them don't have a variable we can test.  The solution that @ryanhaticus proposed seems to work perfectly in those scenarios.

**Reply by ryanhaticus** [reply]
For Twitch, this is the ideal implementation @chris-m-taylor. Great job with the code @itaquito - the `typeof Twitch` check is awesome.
If you'd like, feel free to mark my answer. :)

**Comment by itaquito** [comment]
For the record, this was fixed on #38849
